source/extensions/http/custom_response/redirect_policy/redirect_policy.cc

```diff
@@ -30,23 +57,33 @@ RedirectPolicy::RedirectPolicy(
     const envoy::extensions::http::custom_response::redirect_policy::v3::RedirectPolicy& config,
     Stats::StatName stats_prefix, Envoy::Server::Configuration::ServerFactoryContext& context)
     : stat_names_(context.scope().symbolTable()),
-      stats_(stat_names_, context.scope(), stats_prefix), host_(config.host()),
-      path_(config.path()), status_code_{config.has_status_code()
-                                             ? absl::optional<::Envoy::Http::Code>(
-                                                   static_cast<::Envoy::Http::Code>(
-                                                       config.status_code().value()))
-                                             : absl::optional<::Envoy::Http::Code>{}},
+      stats_(stat_names_, context.scope(), stats_prefix),
+      uri_{config.has_uri() ? std::make_unique<std::string>(config.uri()) : nullptr},
+      redirect_action_{config.has_redirect_action()
+                           ? std::make_unique<::Envoy::Http::Utility::RedirectConfig>(
+                                 createRedirectConfig(config.redirect_action()))
+                           : nullptr},
+      status_code_{config.has_status_code()
+                       ? absl::optional<::Envoy::Http::Code>(
+                             static_cast<::Envoy::Http::Code>(config.status_code().value()))
+                       : absl::optional<::Envoy::Http::Code>{}},
       response_header_parser_(
           Envoy::Router::HeaderParser::configure(config.response_headers_to_add())),
       request_header_parser_(
           Envoy::Router::HeaderParser::configure(config.request_headers_to_add())),
       modify_request_headers_action_(createModifyRequestHeadersAction(config, context)) {
-  ::Envoy::Http::Utility::Url absolute_url;
-  const std::string uri(absl::StrCat(host_, path_));
-  if (!absolute_url.initialize(uri, false)) {
-    throw EnvoyException(
-        absl::StrCat("Invalid uri specified for redirection for custom response: ", uri));
+  if (uri_) {
+    ::Envoy::Http::Utility::Url absolute_url;
+    if (!absolute_url.initialize(*uri_, false)) {
+      throw EnvoyException(
+          absl::StrCat("Invalid uri specified for redirection for custom response: ", *uri_));
+    }
+  }
+  if (redirect_action_ && redirect_action_->host_redirect_.empty() &&
+      redirect_action_->path_redirect_.empty()) {
+    throw EnvoyException("At least one of host_redirect and path_redirect needs to be specified");
   }
+  ASSERT((uri_ || redirect_action_) && !(uri_ && redirect_action_));
```

Comment:
Can you add a comment explaining the invariant?
Should it be at the beginning of the function (not masked by the previous if statements)?